scribus/plugins/import/pdf/slaoutput.cpp

```diff
@@ -3896,3 +3924,670 @@ bool SlaOutputDev::checkClip()
+							,100,0
+							,100,0
+							,100,0
+							,100,100
+							,100,100
+							,100,100
+							,100,100
+							,0,100
+							,0,100
+							,0,100
+							,0,100
+							,0,0
+							,0,0
+	};
+	boundingBoxShape.svgMoveTo(bbosdoubles[0], bbosdoubles[1]);
+	for (int a = 0; a < 16; a += 2)
+	{
+		boundingBoxShape.append(FPoint(bbosdoubles[a * 2], bbosdoubles[a * 2 + 1]));
+	}
+	boundingBoxShape.scale(textNode->width() / 100.0, textNode->height() / 100.0);
+
+	textNode->SetFrameShape(32, bbosdoubles);
+	textNode->ContourLine = textNode->PoLine.copy();
+
+	m_doc->Items->removeLast();
+	m_Elements->append(textNode);
+	if (m_groupStack.count() != 0)
+	{
+		m_groupStack.top().Items.append(textNode);
+		applyMask(textNode);
+	}
+}
+
+/*code mostly taken from importodg.cpp which also supports some line styles and more fill options etc...*/
+//FIXME: This needs to be implemented based on the framework
+void SlaOutputDev::finishItem(PageItem* item)
+{
+	item->ClipEdited = true;
+	item->FrameType = 3;
+	//this requires that PoLine is set
+	//FPoint wh = getMaxClipF(&item->PoLine);
+	//item->setWidthHeight(wh.x(), wh.y());
+	//item->Clip = flattenPath(item->PoLine, item->Segments);
+	item->OldB2 = item->width();
+	item->OldH2 = item->height();
+	item->updateClip();
+	item->OwnPage = m_doc->OnPage(item);
+	//item->setFillTransparency(1.0 - state->getFillOpacity() > state->getStrokeOpacity() ? state->getFillOpacity() : state->getStrokeOpacity());
+	//item->setLineTransparency(1.0);
+}
+
+
+void AddFirstChar::addChar(GfxState* state, double x, double y, double dx, double dy, double originX, double originY, CharCode code, int nBytes, Unicode const* u, int uLen)
+{
+	//qDebug() << "addChar() '" << u << " : " << uLen;	
+	PdfGlyph newGlyph;
+	newGlyph.dx = dx;
+	newGlyph.dy = dy;
+
+	m_textFramework->addChar = m_textFramework->addCharModes[TextFramework::AddCharMode::ADDBASICCHAR];
+
+	// Convert the character to UTF-16 since that's our SVG document's encoding	
+	for (int i = 0; i < uLen; i++)
+	{
+		newGlyph.code = (char16_t)u[i];
+	}
+	
+	newGlyph.rise = state->getRise();
+	//m_slaOutputDev->activeTextRegion.lastXY = QPointF(x, y);
+	m_textFramework->activeTextRegion.glyphs.push_back(newGlyph);
+
+	//only need to be called for the very first point
+	if (m_textFramework->activeTextRegion.addGlyphAtPoint(QPointF(x, y), newGlyph) == TextRegion::FrameworkLineTests::FAIL)
+	{
+		qDebug("FIXME: Rogue glyph detected, this should never happen because the coursor should move before glyphs in new regions are added.");
+	}
+}
+
+void AddBasicChar::addChar(GfxState* state, double x, double y, double dx, double dy, double originX, double originY, CharCode code, int nBytes, Unicode const* u, int uLen)
+{
+	//qDebug() << "addChar() '" << u << " : " << uLen;
+	PdfGlyph newGlyph;
+	newGlyph.dx = dx;
+	newGlyph.dy = dy;
+
+	// Convert the character to UTF-16 since that's our SVG document's encoding
+	for (int i = 0; i < uLen; i++)
+	{
+		newGlyph.code = (char16_t)u[i];

```

Comment:
is it a case for a `static_cast`?

Reply:
and code is always set to the last item in `u`...

Reply:
ok